mlflow/pyspark/optuna/study.py

```diff
@@ -0,0 +1,183 @@
+
+            import optuna
+            import pandas as pd
+
+            import mlflow
+            from mlflow.optuna.storage import MlflowStorage
+
+            mlflow.set_tracking_uri(mlflow_tracking_env)
+
+            storage = MlflowStorage(experiment_id=experiment_id)
+            study = optuna.load_study(study_name=study_name, sampler=sampler, storage=storage)
+            num_trials = sum(map(len, iterator))
+
+            try:
+                error_message = []
+                _optimize_sequential(
+                    study,
+                    func,
+                    num_trials,
+                    timeout,
+                    catch,
+                    callbacks,
+                    False,
+                    False,
+                    datetime.datetime.now(),
+                    None,
+                )
+                error_message.append(None)
+            except BaseException:
+                _traceback_string = traceback.format_exc()
+                error_message.append(_traceback_string)
+                filename = "error_message.txt"
+                with open(filename, "w") as f:
+                    f.write(_traceback_string)
+                # Log the file as an artifact in the active MLflow run
+                mlflow.log_artifact(filename)
+                mlflow.set_tag("LOG_STATUS", "FAILED")
+            finally:
+                df = pd.DataFrame(
+                    {
+                        "error": error_message,
+                    }
+                )
+                yield df
+
+        num_tasks = n_trials
+        if n_jobs == -1:
+            n_jobs = num_tasks
+        input_df = self.spark.range(start=0, end=num_tasks, step=1, numPartitions=n_jobs)
+        trial_tag = f"optuna_trial_{study_name}_{experiment_id}"
+        if self._is_spark_connect_mode:
+            self.spark.addTag(trial_tag)
+        else:
+            job_group_id = self.spark.sparkContext.getLocalProperty("spark.jobGroup.id")
+            if job_group_id is None:
+                job_group_id = trial_tag
+                job_group_description = f"optuna_trial_{study_name}"
+                self.spark.sparkContext.setJobGroup(
+                    job_group_id, job_group_description, interruptOnCancel=True
+                )
+        try:
+            input_df.mapInPandas(
+                func=run_task_on_executor_pd,
+                schema="error string",
+            ).collect()
+        except KeyboardInterrupt as e:
+            if self._is_spark_connect_mode:
+                self.spark.interruptTag(trial_tag)
+            else:
+                self.spark.sparkContext.cancelJobGroup(trial_tag)
+            logger.debug("MLFlowSparkStudy.optimize terminated by user.")
+            raise e
+        if "error" in input_df.columns:
+            error_count = input_df.filter(col("error") != "").count()
+            if error_count > 0:
+                raise ExecutionException("optuna optimize run failed")
```

Comment:
Can we include the first error in the exception message ?

and we can also tell user to see all failures (for every tuning task) in the logged MLflow run.